test/distributed/org/apache/cassandra/distributed/test/DecommissionTest.java

```diff
@@ -0,0 +1,153 @@
+import static org.apache.cassandra.db.SystemKeyspace.BootstrapState.DECOMMISSION_FAILED;
+import static org.apache.cassandra.distributed.api.Feature.GOSSIP;
+import static org.apache.cassandra.distributed.api.Feature.JMX;
+import static org.apache.cassandra.distributed.api.Feature.NATIVE_PROTOCOL;
+import static org.apache.cassandra.distributed.api.Feature.NETWORK;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class DecommissionTest extends TestBaseImpl
+{
+    @Test
+    public void testDecommission() throws Throwable
+    {
+        try (Cluster cluster = init(Cluster.build(2)
+                                           .withConfig(config -> config.with(GOSSIP)
+                                                                       .with(NETWORK)
+                                                                       .with(JMX)
+                                                                       .with(NATIVE_PROTOCOL))
+                                           .withInstanceInitializer(DecommissionTest.BB::install)
+                                           .start()))
+        {
+            IInvokableInstance instance = cluster.get(1);
+
+            assertInfoBootstrapState(instance, COMPLETED);
+
+            instance.runOnInstance(() -> {
+                try
+                {
+                    StorageService.instance.decommission(true);
+                    fail("the first attempt to decommission should fail");
+                }
+                catch (Throwable t)
+                {
+                    assertEquals("Error while decommissioning node: simulated error in prepareUnbootstrapStreaming", t.getMessage());
+                }
+
+                assertTrue("system.local column 'bootstrapped' should contain value " + DECOMMISSION_FAILED,
+                           SystemKeyspace.hasDecommissionFailed());
+
+                assertEquals("system.local column 'bootstrapped' should contain value " + DECOMMISSION_FAILED,
+                             DECOMMISSION_FAILED.name(),
+                             StorageService.instance.getBootstrapState());
+            });
+
+            assertInfoBootstrapState(instance, DECOMMISSION_FAILED);
+
+            instance.runOnInstance(() -> {
+                try
+                {
+                    StorageService.instance.decommission(true);
+
+                    assertFalse("system.local column 'bootstrapped' should not contain value " + DECOMMISSION_FAILED,
+                                SystemKeyspace.hasDecommissionFailed());
+
+                    assertEquals("system.local column 'bootstrapped' should contain value " + DECOMMISSIONED,
+                                 DECOMMISSIONED.name(),
+                                 StorageService.instance.getBootstrapState());
+                }
```

Comment:
should we add a fail("some messages "); after ` assertEquals("system.local column 'bootstrapped' should contain value " + DECOMMISSIONED,
                                 DECOMMISSIONED.name(),
                                 StorageService.instance.getBootstrapState());`
as if the content in this try does not throw an exception?

Reply:
I dont understand. When assertEquals fails, it will throw `ComparisionFailure` / AssertionError. That is caught by `catch` and rethrown?